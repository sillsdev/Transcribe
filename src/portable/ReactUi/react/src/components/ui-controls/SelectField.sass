@import "../../_variables.sass"

.SelectField
    width: 320px
    height: 56px !important
    display: flex
    flex-direction: column
    background: transparent
    justify-content: flex-start
    padding-bottom: 20px

    .bodyRow
        display: flex
        flex-direction: row
        background-color: rgba(255, 255, 255, 0.15)
        padding: 9px 12px 0 12px

    .dataColumn
        display: flex
        flex-direction: column
        flex: 1 0 auto

    .Caption
        font-family: $sans
        font-size: 12px
        height: 16px
        letter-spacing: 0.4px
        line-height: 16px

    .Body
        font-family: $sans
        line-height: 28px
        display: flex
    .DataColor
        color: $base2
    .CaptionGrey
        color: $base1

    select
        flex: 1 0 auto
        border: none
        background: url("/assets/Dropdown.svg") right / 5% no-repeat transparent
        -webkit-appearance: none
        -moz-appearance: none
        font-family: $sans
        line-height: 28px
        font-size: 16px
        letter-spacing: 0.444444px
        &.rtl
            background: url("/assets/Dropdown.svg") left / 5% no-repeat transparent

        option
            margin: 10px 0px
            background: #2a525a
            color: $base2
            border: none
            line-height: 22px
            font-size: 14px

    .Message
        font-family: $sans
        font-size: 12px
        height: 16px
        letter-spacing: 0.4px
        line-height: 16px
        padding: 0 12px

    .BorderTopGreen
        border-top: 2px solid $greenLight

    .BorderTopRed
        border-top: 2px solid $red

    .CaptionGreen
        color: $greenLight

    .CaptionRed
        color: $red
