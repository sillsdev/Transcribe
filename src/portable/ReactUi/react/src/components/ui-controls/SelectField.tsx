import * as React from 'react';
import './SelectField.sass';

interface IProps {
    id?: string;
    caption: string;
    direction?: string;
    options: string[];
    selected?: string;
    message?: string;
    onChange?: (value: string) => any;
}

class TextSelectFieldUx extends React.Component<IProps, any> {
    public state = {
        current: this.props.selected,
    }

    constructor(props: IProps) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
    }

    public handleChange(event: any) {
        /* alert("event.target.value: " + event.target.value)
        alert("this.props.caption: " + this.props.caption) */
        if (event.target.value === this.props.caption) {
            this.setState({ current: undefined });
            if (this.props.onChange != null) {
                this.props.onChange("")
            }
        } else {
            this.setState({ current: event.target.value });
            if (this.props.onChange != null) {
                this.props.onChange(event.target.value)
            }
        }
    }

    public render() {
        const { current } = this.state;
        const { caption, direction, id, message, options } = this.props;
        const errorMessage = (current && current !== "" && message != null)? message: "";
        const messageStyle = (errorMessage.length > 0) ? "Message CaptionRed BorderTopRed" : "Message BorderTopGreen";
        const captionStyle = (errorMessage.length > 0) ? "Caption CaptionRed" : "Caption CaptionGreen";
        const dataStyle = (current && current !== "" && current !== caption)? "DataColor": "CaptionGrey";
        let optionsList;
        if(optionsList === undefined) {
            optionsList = [caption].concat(options)
        }
        const elementList = optionsList.map((item: string) =>
            <option key={this.skey(item)} value={this.skey(item)}>{this.svalue(item)}</option>
        );
        // alert("elementList: " + elementList)

        const captionText = (current && current !== "" && current !== caption)? caption: "";

        return (
            <div id={id} className="SelectField">
                <div className="bodyRow">
                    <div className="dataColumn">
                        <label  className={captionStyle}>{captionText}</label>
                        <select className={"Body " + dataStyle + (direction && direction === "rtl"? " rtl": "")}
                            onChange={this.handleChange}
                            defaultValue={current}>
                                {elementList}
                        </select>
                    </div>
                </div>
                <div className={messageStyle}>
                    <label>{message}</label>
                </div>
            </div>
        )
    }

    private skey(item: string) {
        const pieces = item.split(":")
        return pieces.length > 1? pieces[0]: item;
    }

    private svalue(item: string) {
        const pieces = item.split(":")
        return pieces.length > 1? pieces[1]: item;
    }
}

export default TextSelectFieldUx;