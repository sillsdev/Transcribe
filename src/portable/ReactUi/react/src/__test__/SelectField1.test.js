/* jshint esversion: 6 */
import React from 'react';
import expect from 'expect';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import SelectField from '../components/ui-controls/SelectField';

configure({ adapter: new Adapter() })

// Snapshot for SelectField
describe('>>>Control: SelectField1 --- Snapshot',()=>{
    const minProps = {
        caption: "Caption",
        direction: "ltr",
        id: "SelectField1",
        options: ["OptionA", "OptionB", "OptionC"],
        selected: 0,
    };

    it('+++capturing Snapshot of SelectField1', () => {
        const renderedValue =  renderer.create(<SelectField {...minProps}/>).toJSON()
        expect(renderedValue).toMatchSnapshot(1);
    });
});
//*******************************************************************************************************
describe('>>>Control: SelectField1', () => {
    let wrapper;
    const minProps = {
        caption: "Caption",
        direction: "ltr",
        id: "SelectField1",
        options: ["OptionA", "OptionB", "OptionC"],
        selected: 0,
    };

    beforeEach(() => {
        wrapper = shallow(<SelectField {...minProps} />);
    });

    it('+++ renders without exploding', () => {
        expect(wrapper.length).toEqual(1);
    });
});
