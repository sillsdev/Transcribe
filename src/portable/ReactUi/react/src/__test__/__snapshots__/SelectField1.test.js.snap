// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 1`] = `
<div
  className="SelectField"
  id="SelectField1"
>
  <div
    className="bodyRow"
  >
    <div
      className="dataColumn"
    >
      <label
        className="Caption CaptionGreen"
      >
        
      </label>
      <select
        className="Body CaptionGrey"
        defaultValue={0}
        onChange={[Function]}
      >
        <option
          value="Caption"
        >
          Caption
        </option>
        <option
          value="OptionA"
        >
          OptionA
        </option>
        <option
          value="OptionB"
        >
          OptionB
        </option>
        <option
          value="OptionC"
        >
          OptionC
        </option>
      </select>
    </div>
  </div>
  <div
    className="Message BorderTopGreen"
  >
    <label />
  </div>
</div>
`;
